### PR TITLE
handbook: add observability developer guides

### PR DESCRIPTION
**Pull request by slimsag**
This adds developer documentation for best-practices for developing the observability of Sourcegraph.

Right now, only documentation for monitoring is added. In the future, documentation about how we develop logging and distributed tracing should be added.

This is a first pass, and can certainly be improved in some areas -- but I have had three people ask me questions about this and be confused by the lack of documentation so I want to immediately remediate that.

Once merged, you can find these docs at https://about.sourcegraph.com/handbook/engineering/distribution/observability

### CODE REVIEW

File: handbook/engineering/distribution/observability/monitoring.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+We use [Prometheus](https://prometheus.io) for:
+
+- Collecting high-level, and low-cardinality, metrics from our services.
+- Defining alerting rules (we do not use [Prometheus's Alertmanager](https://prometheus.io/docs/alerting/alertmanager/) but instead define alerts as a single Prometheus metric, more on this later.)
+
+We use [Grafana](https://grafana.com) for:
+
+- Displaying dashboards for our Prometheus metrics.
+- Sending our Prometheus metric alerts to site admins via email, slack, etc.
```
Comment (uwedeportivo) [inline handbook/engineering/distribution/observability/monitoring.md:46]:
```suggestion
- Sending our Prometheus metric alerts to site admins via email, Slack, etc.
```

File: handbook/engineering/distribution/observability/monitoring.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+- Generating the Prometheus alerting rules and Grafana dashboards.
+- Generating documentation in the form of ["possible solutions"](https://docs.sourcegraph.com/admin/observability/alert_solutions) for site admins to follow when alerts are firing.
+
+## The five pillars of monitoring
+
+At Sourcegraph we impose five strict constraints to monitoring which, at first, may seem quite surprising:
+
+1. Creating dashboards that describe something other than a single Sourcegraph service is forbidden.
+2. Adding a graph/panel that visualizes a metric, without defining an associated alert, is forbidden.
+3. Creating dashboards outside of the monitoring generator, such as with the Grafana WYSIWYG editor, is forbidden. All metrics should be presented in the most plain, mundane, non-fanciful way that every site admin can understand.4. Creating graphs/panels with more than 5 cardinality (labels) is forbidden (in most cases there should only be one.)
```
Comment (daxmc99) [inline handbook/engineering/distribution/observability/monitoring.md:63]:
```suggestion
3. Creating dashboards outside of the monitoring generator, such as with the Grafana WYSIWYG editor, is forbidden. All metrics should be presented in the most plain, mundane, non-fanciful way that every site admin can understand.
4. Creating graphs/panels with more than 5 cardinalities (labels) is forbidden (in most cases there should only be one.)
```

File: handbook/engineering/distribution/observability/monitoring.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
```
Comment (uwedeportivo) [inline handbook/engineering/distribution/observability/monitoring.md]:
```suggestion
3. Creating dashboards outside of the monitoring generator, such as with the Grafana WYSIWYG editor, is forbidden. All metrics should be presented in the most plain, mundane, non-fanciful way that every site admin can understand.
4. Creating graphs/panels with more than 5 labels is forbidden (in most cases there should only be one.)
```

File: handbook/engineering/distribution/observability/monitoring.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
++	PossibleSolutions: `
++       - Look at the 'frontend' logs for details on the slow search queries.
++   `,
+}
+```
+
+> Tip: In `PossibleSolutions` Markdown, you can use single quotes anywhere you would normally use backticks for code, and indention will automatically be removed for you.
+
+
+One you save the file, `doc/admin/observability/alert_solutions.md` will automatically be regenerated and you can even preview your changes at [http://localhost:5080/admin/observability/alert_solutions](http://localhost:5080/admin/observability/alert_solutions).
```
Comment (uwedeportivo) [inline handbook/engineering/distribution/observability/monitoring.md:128]:
```suggestion
Once you save the file, `doc/admin/observability/alert_solutions.md` will automatically be regenerated and you can even preview your changes at [http://localhost:5080/admin/observability/alert_solutions](http://localhost:5080/admin/observability/alert_solutions).
```

**Comment by slimsag** [comment]
Confirmed `$ make docsite-check` passing locally, Netlify failure seems unrelated

File: handbook/engineering/distribution/observability/monitoring_pillars.md [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+4. Creating graphs/panels with more than 5 cardinality (labels) is forbidden (in most cases there should only be one.)
+5. The most useful information should be presented first, the least useful information should be hidden by default.
+
+To understand why we impose these constraints, see [the five pillars of monitoring](monitoring_pillars.md)
+
+## Monitoring at Sourcegraph: a brief history
+
+Before the constraints we impose will start to make sense, you need to understand the history of monitoring at Sourcegraph and the (extremely difficult) problem we're trying to solve with it.
+
+### For as long as Sourcegraph has existed as an enterprise product, it has had some form of Prometheus and Grafana monitoring
```
Comment (uwedeportivo) [inline handbook/engineering/distribution/observability/monitoring_pillars.md:34]: I would not include this history

File: handbook/engineering/distribution/observability/monitoring_pillars.md [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+You might be thinking:
+
+> But writing code for dashboards is quite painful, let alone a bunch of generation code that I now need to understand!
+
+In truth, we are only using Go as a nice type-checked declarative syntax. It's as simple to use as a configuration file, and the constraints we impose in monitoring means that we do not allow doing any fanciful dashboards intentionally -- which means you are basically _just_ writing a Prometheus query, and barely any layout or styling.
+
+See: [How easy is it to add monitoring?](monitoring.md#how-easy-is-it-to-add-monitoring)
+
+### FAQ: Why can't I create a graph/panel with more than 5 cardinality (labels)?
```
Comment (daxmc99) [inline handbook/engineering/distribution/observability/monitoring_pillars.md:139]: Hmm.... I feel like these conflicts with other observability guidelines https://thenewstack.io/observability-a-3-year-retrospective/
From the article: 
> For example, observability is impossible without:

    raw events
    high cardinality dimensions
    no pre-aggregation, no pre-indexing (which lock you into asking predefined questions)
    read time aggregation

File: handbook/engineering/distribution/observability/index.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Sourcegraph observability developer guide
+
+**Note:** Looking for _how to monitor Sourcegraph?_ See the [observability documentation](https://docs.sourcegraph.com/admin/observability).
+
+We use three tools primarily to observe (monitor and debug) Sourcegraph:
+
+- [Monitoring](monitoring.md) 
+- Logging (developer docs coming soon, for now see [admin docs](https://docs.sourcegraph.com/admin/observability))
+- Distributed tracing (developer docs coming soon, for now see [admin docs](https://docs.sourcegraph.com/admin/observability))
```
Comment (daxmc99) [inline handbook/engineering/distribution/observability/index.md:9]: Excited to add to this in the future!

File: handbook/engineering/distribution/observability/monitoring_pillars.md (hunk 6 above)
Comment (daxmc99) [reply handbook/engineering/distribution/observability/monitoring_pillars.md:124]: On a second read-through, I think this is fine for _monitoring_. 
A section on the difference between monitoring and observability would be welcome though. 
These are terms that are still evolving though

Comment (slimsag) [reply handbook/engineering/distribution/observability/monitoring_pillars.md:124]: I do believe your quote is true!

The difference is that:

1. We get our raw events and cardinality from other tools that are better at handling that cardinality (e.g. logging, distributed tracing)
2. Monitoring is just one part of our observability stack, and in cases where higher cardinality in monitoring makes sense we _do allow it_ just not in individual graphs/panels (you need to split it into multiple panels/graphs for improved visibility).

File: handbook/engineering/distribution/observability/monitoring_pillars.md (hunk 5 above)
Comment (slimsag) [reply handbook/engineering/distribution/observability/monitoring_pillars.md]: Shortened it substantially

**Comment by slimsag** [comment]
Merging so I can link others on the team who have an immediate need for this, and to get more feedback. Will follow up on all feedback (none of this is set in stone, just what we're doing today)

File: handbook/engineering/distribution/observability/monitoring_pillars.md (hunk 6 above)
Comment (slimsag) [reply handbook/engineering/distribution/observability/monitoring_pillars.md:124]: > A section on the difference between monitoring and observability would be welcome though.

Yes, definitely, I have an issue open here for me to add this: https://github.com/sourcegraph/sourcegraph/issues/9773

Comment (slimsag) [reply handbook/engineering/distribution/observability/monitoring_pillars.md:124]: Large part done in https://github.com/sourcegraph/about/pull/948

**Pull request merged**